### PR TITLE
update Op concept doc

### PR DESCRIPTION
**Pull request by smackesey**
## Summary

This is a significant edit of the first half (Relevant APIs and Overview) of the Op concept guide. This is more thorough then I had been initially thinking, but IMO the result is a significant improvement. I think many of the Concept pages would benefit from similar treatment.

### CODE REVIEW

**Comment by schrockn** [comment]
These changes are excellent 👍🏻

**Comment by smackesey** [comment]
@schrockn Thanks. What do you think about massaging the topics of these pages:

- Ops [RENAME to Ops & Graphs]
- Jobs & Graphs [RENAME to Jobs & Runs]
- Job Execution [MERGE into Jobs & Runs]

Why?

- Graphs fit more naturally with Ops than Jobs. Ops/Graphs specify computations but aren't executable. Jobs are executable. Graphs are also confusingly presented right now, they are always grouped with Jobs but then barely mentioned.
- "Runs" are an important concept (there is a dedicated section for them in dagit), but get short shrift in the docs.
- Job Execution page is pretty thin and being executable is the defining feature of jobs anyway (when contrasted with graphs).

File: docs/content/concepts/ops-jobs-graphs/ops.mdx [hunk 1]
```diff
@@ -1,42 +1,41 @@
...
+The op as computational unit enables many useful features for data orchestration:
 
-Ops have several important properties:
-
-- **Inputs and Outputs**: Ops have defined [inputs and outputs](#inputs-and-outputs), which can be optionally typed. These types are validated at runtime.
-- **Configurable**: Ops can be configured, using a strongly typed [configuration system](/concepts/configuration/config-schema).
-- **Emit an Event Stream**: Ops can emit a stream of structured events, such as AssetMaterializations. These events can be viewed in [Dagit](/concepts/dagit/dagit), Dagster's UI tool.
-- **Individually testable**: See the [testing page](/concepts/testing) for more detail.
+- **Flexible execution strategies**: Data pipelines are frequently developed locally and later deployed to production. Ops represent sealed units of logic independent of pipeline execution strategy, making the transition from development to production painless. Collections of ops (i.e. [Jobs](/concepts/ops-jobs-graphs/jobs-graphs)) can be passed to an appropriate [Executor](/deployment/executors) for single-process execution or distribution across a cluster.
```
Comment (sryza) [inline docs/content/concepts/ops-jobs-graphs/ops.mdx:34]: "passed to an appropriate Executor" isn't exactly accurate. Each job has an executor - i.e. the executor is determined at job definition time.

File: docs/content/concepts/ops-jobs-graphs/ops.mdx [hunk 2]
```diff
@@ -1,42 +1,41 @@
...
+- **Input and output management**: Ops have defined [inputs and outputs](#inputs-and-outputs) analogous to the arguments and return value(s) of a Python function. However, op inputs/outputs can be more richly annotated than Python function arguments. This annotation can instruct Dagster's engines how to validate, cache, serialize, and otherwise manage data as it flows between ops.
```
Comment (sryza) [inline docs/content/concepts/ops-jobs-graphs/ops.mdx:35]: I think replacing the last sentence with something more specific could make this bullet more clear.  E.g. something like "Annotating inputs/outputs with Dagster Types instructs Dagster on how to validate them, and annotating them with io_manager_keys instructs Dagster on how to serialize them." Would also omit "engines" to avoid risking users thinking that an "engine" is a meaningful abstraction in Dagster.

File: docs/content/concepts/ops-jobs-graphs/ops.mdx [hunk 3]
```diff
@@ -1,42 +1,41 @@
...
+- **Configuration**: Operations in a data pipeline are often parameterized by both upstream data (e.g. a stream of database records) and "configuration" values independent of upstream data (e.g. a "chunk size" of incoming records to operate on). Op configuration values can be defined and annotated in much the same way as inputs and outputs.
+- **Event streams**: Ops emit a stream of [Events](/concepts/ops-jobs-graphs/op-events) during execution. Certain events are emitted by default (e.g. indicating the start of an op's execution), but op authors are additionally given access to an event API. This can be used to report data asset creation or modification (<PyObject object="AssetMaterialization"/>), the result of a data quality check (<PyObject object="ExpectationResult"/>), or other arbitrary information. Event streams can be visualized by Dagster's browser UI [Dagit](/concepts/dagit/dagit). This rich log of execution facilitates debugging, inspection, and real-time monitoring of pipeline execution and data asset production.
```
Comment (sryza) [inline docs/content/concepts/ops-jobs-graphs/ops.mdx:37]: "execution" twice in the last sentence is a tiny bit awkward. Could just leave out the words after "monitoring"?

File: docs/content/concepts/ops-jobs-graphs/ops.mdx [hunk 4]
```diff
@@ -1,42 +1,41 @@
...
+| <PyObject object="OpExecutionContext"/> | An object exposing Dagster system APIs for resource access, logging, and more. Can be injected into an op by specifying `context` as the first argument of the compute function.                           |
+| <PyObject object="OpDefinition"  />     | Class for ops. You will rarely want to instantiate this class directly. Instead, you should use the <PyObject object="op" decorator />. |
 
 ## Overview
 
-Ops are used to define computations. Ops can later be assembled into [Jobs](/concepts/ops-jobs-graphs/jobs-graphs). Ops generally perform one specific action and are used for batch computations. For example, you can use an op to:
+Ops are Dagster's atomic unit of computation. Individual ops should perform relatively simple tasks. Collections of ops can then be assembled into [Jobs or Graphs](/concepts/ops-jobs-graphs/jobs-graphs) to perform more complex tasks. Some examples of tasks appropriate for a single op:
 
-- Derive a data set from some other data sets.
```
Comment (sryza) [inline docs/content/concepts/ops-jobs-graphs/ops.mdx:26]: Any particular reason to take this one out? IMO, this is the top task for an op.

**Review by sryza — CHANGES_REQUESTED** [review]
I think this is a great improvement. I commented on a few individual points.

File: docs/content/concepts/ops-jobs-graphs/ops.mdx (hunk 4 above)
Comment (smackesey) [reply docs/content/concepts/ops-jobs-graphs/ops.mdx:26]: My reason for this change is that "derive a dataset from another dataset" is vague wrt scale. This could describe an op, or an entire pipeline, or something even larger.

File: docs/content/concepts/ops-jobs-graphs/ops.mdx (hunk 3 above)
Comment (smackesey) [reply docs/content/concepts/ops-jobs-graphs/ops.mdx:37]: Good catch

File: docs/content/concepts/ops-jobs-graphs/ops.mdx [hunk 5]
```diff
@@ -1,42 +1,41 @@
 ---
 title: Ops | Dagster
-description: Ops are the functional unit of computation in Dagster and contain the core logic of your orchestration graph.
+description: Ops are the atomic unit of computation in Dagster and contain the core logic of your orchestration graph.
```
Comment (schrockn) [inline docs/content/concepts/ops-jobs-graphs/ops.mdx:3]: Do we like the word "atomic"? It might imply transactionality that we don't actually have.

Comment (yuhan) [reply docs/content/concepts/ops-jobs-graphs/ops.mdx:3]: I have a bit reservations on the work "atomic" - given that many of our users (and prospective users) are familiar with DB systems, it could imply atomic transaction and therefore cause confusion.

File: docs/content/concepts/ops-jobs-graphs/ops.mdx (hunk 4 above)
Comment (yuhan) [reply docs/content/concepts/ops-jobs-graphs/ops.mdx:26]: I like keeping this one. Agreed with Sean's assessment but because of the same reason, I like the ambiguity here to show that an op the where you process a data set.

File: docs/content/concepts/ops-jobs-graphs/ops.mdx (hunk 2 above)
Comment (smackesey) [reply docs/content/concepts/ops-jobs-graphs/ops.mdx:35]: Will adjust last sentence.

Thx for the pointer on "engine", I can delete that. But this brings up a point I'm hazy on-- is there a central entity that coordinates job execution, resource access, etc? What is it called? I have seen a few places where different terms are used:

- "An optional first argument to a op's compute function, providing access to Dagster system information such as resources, logging, and the underlying **Dagster instance**."
- "Yielding events from within the body of an op is a useful way of communicating with the **Dagster framework**."
- "It is also possible to raise Dagster-specific exceptions, which will indicate to the **framework** to halt the op execution entirely and perform some action."
-"The **Dagster Daemon** operates schedules, sensors, and run queuing. Currently, replicas are not supported." 
- "ENGINE EVENT" is a DagsterEventType that shows up in Dagit.

"Framework" here seems wrong, as it is not a specific entity that communicates with other entities. My guess is "Dagster Instance" is best (does this mean an instance of the Daemon?), but this term is barely used in the docs.

File: docs/content/concepts/ops-jobs-graphs/ops.mdx (hunk 1 above)
Comment (yuhan) [reply docs/content/concepts/ops-jobs-graphs/ops.mdx:34]: I like the explanation here! Maybe rephrase this bullet "Flexible execution strategies" to "Business logic separation" or "Environment isolation"? bc besides executors, an op can also isolate the biz logic from resources across different execution environments.

**Review by yuhan — COMMENTED** [review]
Thank you for pushing this forward. I think the changes are making this page from a tech concept deep dive to an understanding-oriented detailed explanation of why things are so.

File: docs/content/concepts/ops-jobs-graphs/ops.mdx (hunk 2 above)
Comment (yuhan) [reply docs/content/concepts/ops-jobs-graphs/ops.mdx:35]: Dagster Instance is a specific concept in Dagster: https://docs.dagster.io/deployment/dagster-instance
In this case, I usually would say "This annotation can instruct Dagster **machinery** how to validate, cache, serialize, and otherwise manage data as it flows between ops." - "machinery" is the word we use in code comments and docstring for this case.

File: docs/content/concepts/ops-jobs-graphs/ops.mdx (hunk 1 above)
Comment (smackesey) [reply docs/content/concepts/ops-jobs-graphs/ops.mdx:34]: Thx for pointer on executors. My preferred solution here relates to my point above about graphs.

In several places I've seen collections of ops described as jobs, or references to connecting ops to form a job. For instance, the headlines for Op and Job concept pages in current docs say:

- "Multiple ops can be connected to create a Job."
- "A job is an executable graph of ops"

I understand a decision was made to emphasize jobs over graphs at some point, presumably because it introduces fewer abstractions when you're trying to explain how to get something done.

But I think this has resulted in elisions that are ultimately costly to the user, because it creates confusion about Dagster works. IMO it is better to always describe collections of connected ops  as capital-g "Graphs", and emphasize that a Graph is a component of a Job-- Ops are computational units, they can be wired together into Graphs, and a Job binds configuration/resources and an executor (and maybe some other stuff) with a Graph. Use of the @job decorator is just a way of creating an anonymous Graph, analogous to an inline lambda in Python.

Following this approach, I'd edit that last sentence like so:

>Collections of ops (i.e. [Graphs](/concepts/ops-jobs-graphs/jobs-graphs)) are bound via [Jobs](/concepts/ops-jobs-graphs/jobs-graphs) to an appropriate [Executor](/deployment/executors) for single-process execution or distribution across a cluster.

File: docs/content/concepts/ops-jobs-graphs/ops.mdx (hunk 5 above)
Comment (smackesey) [reply docs/content/concepts/ops-jobs-graphs/ops.mdx:3]: Good point re "atomic"-- I went with this term to imply "no smaller unit of composition". I changed it bc I didn't really understand what "functional" meant. Some alternatives modifiers for "unit" are "core" or "basic", or nothing at all. IMO these communicate less than "atomic", but also avoid the transactionality baggage.

File: docs/content/concepts/ops-jobs-graphs/ops.mdx (hunk 2 above)
Comment (smackesey) [reply docs/content/concepts/ops-jobs-graphs/ops.mdx:35]: Good suggestion on "machinery". And looks like I missed that Instance page, I think because I haven't paid enough attention to "Deployment" section.

File: docs/content/concepts/ops-jobs-graphs/ops.mdx (hunk 5 above)
Comment (yuhan) [reply docs/content/concepts/ops-jobs-graphs/ops.mdx:3]: "core" works for me.

**Comment by sryza** [comment]
> Ops [RENAME to Ops & Graphs]

I wrote an earlier version of our 0.13.0 docs separated jobs from graphs, but I found that putting them together worked better - the concepts felt difficult to explain without each other. When they were separate, I ended up needing to repeat many of the same explanations in both docs. Very possible that someone else could do a better job than me, but IMO it's not worth the effort at this time.

> Job Execution [MERGE into Jobs & Runs]

Agreed that this page is thin. It could make sense to merge it into the jobs & graphs page, even if we don't separate jobs from graphs.

This also makes me think that it might make sense to merge the Op Events page into the Ops page.

**Comment by yuhan** [comment]
Re: https://github.com/dagster-io/dagster/pull/5477#issuecomment-955837100
I'm on board with merging "Job Execution" into "Jobs&Graphs" and merging "Op Events" into "Ops".

File: docs/content/concepts/ops-jobs-graphs/ops.mdx [hunk 6]
```diff
@@ -1,42 +1,42 @@
...
+- **Input and output management**: Ops have defined [inputs and outputs](#inputs-and-outputs) analogous to the arguments and return value(s) of a Python function. An input or output can be annotated with a [Dagster Type](/concepts/types) for arbitrarily complex runtime validation. Outputs can additionally be tagged with an [IO Manager](/concepts/io-management/io-managers) to manage storage of the associated data in between ops. This enables easy swapping of I/O strategy depending on the execution environment, as well as efficient caching of data intermediates (i.e. [memoization](/guides/dagster/memoization)).
```
Comment (sryza) [inline docs/content/concepts/ops-jobs-graphs/ops.mdx:36]: I think need comma after analogous.

Comment (sryza) [reply docs/content/concepts/ops-jobs-graphs/ops.mdx:36]: I think we should avoid fore-fronting memoization here, because it's still experimental (could just leave out the parenthetical?).

**Review by sryza — APPROVED** [review]
I left two small comments - otherwise, I think this looks great.

**Comment by sryza** [comment]
Looks good - feel free to go ahead and merge.
